File: src/main/frontend/src/lib/messaging/model/Message.ts
import {Moment} from "moment";
import {MessageGroup} from "./MessageGroup";
import Messageable from "./Messageable";
import Attachment from "./Attachment";
import MessageSource from "./MessageSource";
import MessagingError from "../../error/MessagingError";

export default class Message
{
	protected _id: string;
	protected _conversationId: string;
	protected _subject: string;
	protected _message: string;
	protected _group: MessageGroup;
	protected _read: boolean;
	protected _sender: Messageable;
	protected _recipients: Messageable[];
	protected _metaData: any;
	protected _createdAtDateTime: Moment;
	protected _attachments: Attachment[];
	protected _source: MessageSource;

	// ==========================================================================
	// Public Methods
	// ==========================================================================

	/**
	 * build new message
	 * @param id - message id
	 * @param conversationId - conversation id to which the message is associated
	 * @param subject - subject of the message
	 * @param message - actual message of the message
	 * @param group - group of the message
	 * @param read - if the message is read or not
	 * @param sender - the sender of the message
	 * @param recipients - recipient list of the message
	 * @param metaData - meta data for the message
	 * @param createdAtDateTime - time at which the message was created.
	 * @param attachments - attachments for this message
	 */
	constructor(
		id: string,
		conversationId: string,
		subject: string,
		message: string,
		group: MessageGroup,
		read: boolean,
		sender: Messageable,
		recipients: Messageable[],
		metaData: any,
		createdAtDateTime: Moment,
		attachments: Attachment[]
	)
	{
		this._id = id;
		this._conversationId = conversationId;
		this._subject = subject;
		this._message = message;
		this._group = group;
		this._read = read;
		this._sender = sender;
		this._recipients = recipients;
		this._metaData = metaData;
		this._createdAtDateTime = createdAtDateTime;
		this._attachments = attachments;
	}

	/**
	 * archive the message
	 */
	public archive(): void
	{
		this._group = MessageGroup.Archived;
	}

	/**
	 * unarchive the message
	 */
	public unarchive(): void
	{
		throw new MessagingError("Base message cannot be unarchived");
	}

	// ==========================================================================
	// Setters
	// ==========================================================================

	set source(messageSource: MessageSource)
	{
		this._source = messageSource;
	}

	set read(read: boolean)
	{
		this._read = read;
	}

	// ==========================================================================
	// Getters
	// ==========================================================================

	get id(): string
	{
		return this._id;
	}

	get conversationId(): string
	{
		return this._conversationId;
	}

	get subject(): string
	{
		return this._subject;
	}

	get message(): string
	{
		return this._message;
	}

	get group(): MessageGroup
	{
		return this._group;
	}

	get read(): boolean
	{
		return  this._read;
	}

	get isRead(): boolean
	{
		return this._read;
	}

	get sender(): Messageable
	{
		return this._sender;
	}

	get recipients(): Messageable[]
	{
		return this._recipients;
	}

	get metaData(): any
	{
		return this._metaData;
	}

	get createdAtDateTime(): Moment
	{
		return this._createdAtDateTime;
	}

	get attachments(): Attachment[]
	{
		return this._attachments;
	}
	
	get hasAttachments(): boolean
	{
		return this._attachments && this._attachments.length > 0;
	}

	get source(): MessageSource
	{
		return this._source;
	}
}

File: src/main/java/org/oscarehr/messaging/model/Message.java
/**
 * Copyright (c) 2012-2018. CloudPractice Inc. All Rights Reserved.
 * This software is published under the GPL GNU General Public License.
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * <p>
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * <p>
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 * <p>
 * This software was written for
 * CloudPractice Inc.
 * Victoria, British Columbia
 * Canada
 */

package org.oscarehr.messaging.model;

import java.time.ZonedDateTime;
import java.util.List;

public interface Message
{
	// ==========================================================================
	// Getters
	// ==========================================================================

	public String getId();
	public String getConversationId();

	public String getSubject();
	public String getMessage();

	/**
	 * get the group of the message
	 * @return - the messages group
	 */
	public MessageGroup getGroup();

	/**
	 * indicate if the user has read the message or not.
	 * @return - true / false
	 */
	public Boolean isRead();

	/**
	 * get the time at which the message was created.
	 * @return - message creation time.
	 */
	public ZonedDateTime getCreatedAtDateTime();

	/**
	 * get the messageable that sent this message.
	 * @return - the sender.
	 */
	public Messageable<?> getSender();

	/**
	 * get list of recipients of this message.
	 * @return - list of recipients.
	 */
	public List<? extends Messageable<?>> getRecipients();

	/**
	 * get JSON string of meta data.
	 * @return - json meta data.
	 */
	public String getMetaData();

	/**
	 * get a list of attachments
	 * @return - the attachments
	 */
	public List<? extends Attachment> getAttachments();

}
